Record git project name and sha

ForgetMe17 opened the pull request:
This pr adds functions to the dockerfile parser. If the parser
comes across the ADD directive, record the git projeect name and
sha in a comment after the command.

This pr modifies utils/general.py and analyze/docker/dockerfile.py.

In dockerfile.py, function find_git_info recieve a line of ADD
command and the path to the parsed dockerfile. Function
expand_add_command receives a dfobj and iter it with find_git_info.

In general.py, function check_git_folder is used to find git info
in a folder and funtion check_git_tar in a tar file. Function
check_git_src looks into the src_path, and check the file type.

Work towards #509.

ForgetMe17 commented:
I have run some local tests. i.e.
```
FROM ubuntu
ADD plain_file /tmp
ADD test_git_repo /tmp
ADD test_git_repo.tar /tmp
```
i got the output 
```
FROM ubuntu
ADD plain_file /tmp # Not a folder or tar file
ADD test_git_repo /tmp # git project name: (not found) sha info: 55cbf5cbc8b201a99ac63f70f6f6f0ff1d1ae630
ADD test_git_repo.tar /tmp # git project name: (not found) sha info: 55cbf5cbc8b201a99ac63f70f6f6f0ff1d1ae630
```
The dockerfile and git repo are in the sample folder.
I am now working on  adding some tests for these functions.

rnjudge commented on tern/analyze/docker/dockerfile.py:
```diff
@@ -265,3 +265,38 @@ def get_base_image_tag(base_instructions):
+    if args[1].startswith('--chown'):
+        # check if the line is valid
+        if len(args) < 4:
+            raise ValueError('Invalid ADD command line')
```
Instead of raising a `ValueError` can you log the error? If we raise the error it will cause Tern to quit and ideally we would continue with the rest of the potential lines in the Dockerfile. Something like:

```
logger.error("Invalid ADD command line: %s", line)
```

rnjudge commented on tern/analyze/docker/dockerfile.py:
```diff
@@ -265,3 +265,38 @@ def get_base_image_tag(base_instructions):
+    else:
+        # the line has no --chown option
+        if len(args) < 3:
+            raise ValueError('Invalid ADD command line')
```
Same comment as above. Log the error instead of raising it.

rnjudge commented on tern/analyze/docker/dockerfile.py:
```diff
@@ -265,3 +265,38 @@ def get_base_image_tag(base_instructions):
+
+
+def expand_add_command(dfobj):
+    dockerfile_path = dfobj.filepath
```
I'm not sure if `dfobj.filepath` will necessarily be an absolute path. @nishakm is it possible that this path could be relative? From what I can tell, `dfobj.filepath` gets assigned to whatever the user specifies as the Dockerfile, which may not be absolute.

ForgetMe17 replied:
I use open() and tarfile.open() to read the src file of ADD command. So once we make sure the relative path is in the sys.path, we can also get the src file.

nishakm replied:
@rnjudge No it's not an absolute path and it doesn't need to be that. We just need to find the git repo and commit HEAD. I don't think we need to read what is in the tar file.

rnjudge commented on tern/utils/general.py:
```diff
@@ -116,6 +117,102 @@ def get_git_rev_or_version():
+
+
+def check_git_tar(path_to_tar):
+    '''Given the path to the tar file, return the name and sha of repository.
```
small nit, but might want to add in this description that if the tarfile is **not** a git repository, this function will also tell you that.

rnjudge commented on tern/utils/general.py:
```diff
@@ -116,6 +117,102 @@ def get_git_rev_or_version():
+    # get the folder path
+    dockerfile_folder_path = os.path.dirname(dockerfile_path)
+    # get the absolute path of the target folder or tar file
+    src_path = os.path.join(dockerfile_folder_path, src_path)
```
What if the potential `git` directory, specified by `src_path`, is not in the same directory as the Dockerfile folder path? Or what if  the Dockerfile is in a few subfolders below? @nishakm not sure how hard we want to look for the source that is being copied in. It's also common to see the source being copied in as just `.` which could be multiple tars or directories. We should probably account for this in some way even if we don't filter through all the directories for potential git information.

ForgetMe17 replied:
In general the src file path is the relative path to the Dockerfile path. I tested that it works with the git repo in the subfolders. If the dockerfile is in a few subfolders below, the given dockerfile_path is changed. What we do here is looking around the folder containing dockerfile and its subfolders. So we need to locate the dockerfile in this situation. I am wondering if I have missed some situations.

rnjudge replied:
>> In general the src file path is the relative path to the Dockerfile path

This is true in most cases, but we need to account for when it is not true. People do weird things with Dockerfiles :) 

>> If the dockerfile is in a few subfolders below, the given dockerfile_path is changed.

If the Dockerfile is a few subfolders below the git folder then wouldn't `src_path = os.path.join(dockerfile_folder_path, src_path)` give you `dir/dockerfile_folder_below/gitdirectory` even though the git directory could be a directory above in `dir`?

rnjudge commented:
> I have run some local tests. i.e.
> 
> ```
> FROM ubuntu
> ADD plain_file /tmp
> ADD test_git_repo /tmp
> ADD test_git_repo.tar /tmp
> ```
> 
> i got the output
> 
> ```
> FROM ubuntu
> ADD plain_file /tmp # Not a folder or tar file
> ADD test_git_repo /tmp # git project name: (not found) sha info: 55cbf5cbc8b201a99ac63f70f6f6f0ff1d1ae630
> ADD test_git_repo.tar /tmp # git project name: (not found) sha info: 55cbf5cbc8b201a99ac63f70f6f6f0ff1d1ae630
> ```
> 
> The dockerfile and git repo are in the same folder.
> I am now working on adding some tests for these functions.

@ForgetMe17 what is the test_git_repo that you have in your test dockerfile? Could you also test with a git repository where the project name is available?

@nishakm Is the sha info useful if we don't know the name of the git project? Also, dumb question but how is this possible? This seems confusing to me but I am also no expert at what .git metadata provides.

ForgetMe17 commented:
I created a local git repository, not cloning from github. So the repo does not have a remote url where I get the name of project.

rnjudge commented:
@nishakm I am thinking we may also want to move `find_git_info`, `check_git_folder`, `check_git_tar` and `check_git_src` to `tern/analyze/docker/common.py` instead of leaving them in `dockerfile.py` and just leave the `expand_add_command` in dockerfile.py. Thoughts?

ForgetMe17 commented:
In the commit `87b35d6`, I add tests for find_git_info and expand_add_command, and move `check_git_folder`, `check_git_tar` and `check_git_src` to `tern/analyze/docker/common.py`.

I test the plain file, tar file and git folder. For git folder, since a git repo cannot be uploaded unless as a submodule. So in the online ci test, the result for the git folder is `Not a git repository`.

In the test, the path of dockerfile is relative and the src_path is also relative. For the situations where the git repo is a few folders above the dockerfile, i think in this case the src_path in the ADD command should be like `../../git_repo`. I donnot know if it is valid in the dockerfile(probably not i think), but this pr does not work with this situation since the `os.path.join` will append `../../git_repo` to the dockerfile_path. So maybe i will work on locating src_file in this situation.

nishakm commented on tern/analyze/common.py:
```diff
@@ -452,3 +454,100 @@ def get_os_style(image_layer, binary):
+    return comment_line
+
+
+def check_git_tar(path_to_tar):
```
I don't think we need to look into the tarball that gets included into the docker image. It is already in there by the time we analyze it. I think this function can be deleted.

nishakm commented on tests/test_analyze_docker_dockerfile.py:
```diff
@@ -102,6 +105,43 @@ def testParseFromImage(self):
+                     'content': ('ADD test_git_repo.tar /tmp # git project '
+                                 'name: test_repo.git sha info: 55cbf5cbc8b'
+                                 '201a99ac63f70f6f6f0ff1d1ae630\n'),
+                     'value': 'test_git_repo.tar /tmp'}]
```
I don't think this functionality is testable, because the git sha of the tern repo will change every time a PR is merged. We'll just have to test the parts that adds the comment and the function that gets the git repository name (if those two things are in different functions).

rnjudge commented on tern/analyze/common.py:
```diff
@@ -452,3 +454,100 @@ def get_os_style(image_layer, binary):
+                sha_path = dot_git_folder_path + ref_relative_path
+                # get sha info from ref
+                with open(sha_path, 'r') as sha_file:
+                    sha_info = sha_file.readline().strip('\n')
```
You can get the sha info in one line using `git rev-parse HEAD`. Take a look at `get_git_rev_or_version()` function in  tern/tern/utils/general.py for an example on how to do this. You may also want to break this into a mini-function called something like `get_git_sha(path)` for clarity.

rnjudge commented on tern/analyze/common.py:
```diff
@@ -452,3 +454,100 @@ def get_os_style(image_layer, binary):
+    comment_line = ''
+    if os.path.isdir(path_to_folder):
+        # find .git folder
+        dot_git_folder_path = os.path.join(path_to_folder, '.git/')
```
You can find the top level directory (that will container .git inside it) with `git rev-parse --show-toplevel` command.

ForgetMe17 replied:
Thanks for your suggestion!

rnjudge commented on tern/analyze/common.py:
```diff
@@ -452,3 +454,100 @@ def get_os_style(image_layer, binary):
+            if conf.has_section('remote "origin"'):
+                # url is the first item
+                project_url = conf.get('remote "origin"', 'url')
+                name_info = project_url.split('/')[-1]
```
This is tricky because while most remotes will be named `origin`, it is possible to rename `origin` to something else. We also want to get the complete repo name (i.e. github.com/tern.git)

Assuming you are in a git repository, `basename `git rev-parse --show-toplevel` will give you just the name of the actual project. Finding the url will be a little harder.

ForgetMe17 commented:
commit `648427f`, Using command line and modified tests
In `tern/analyze/common.py`:
1.Delete function check_git_tar and check_git_folder.
2.Add get_git_toplevel which uses git command to find the toplevel directory.
3.Add get_git_project_name and get_git_sha to get the git repository information which uses the path_to_toplevel given by the get_git_toplevel.

Modified tests in the `tests/test_analyze_docker_dockerfile.py`

In the test, i am using the dockerfile in `tests/dockerfiles/pin_add_command_test` and the ADD command is `ADD plain_file /tmp`. The `plain_file` is also in `tests/dockerfiles/pin_add_command_test`.  `plain_file` is not a git repo, but it is in our `tern` git repository, so the project name we get is `tern`, which is the toplevel folder name of the git repository. But in the circleci test, i find that the project name is `project`. I think it is becasue in the circleci test environment, the toplevel folder name is change to `project`. In the Pull Request Lint and Test of the Github, the toplevel folder name is 'tern'. So in the test, i used AssertIn to match the name between the `tern` and `project`.

And thanks for your suggestions which really helps me a lot!

rnjudge commented on tern/analyze/common.py:
```diff
@@ -452,3 +453,76 @@ def get_os_style(image_layer, binary):
+
+
+def get_git_toplevel(src_path):
+    '''Give a path, return true if it is in a git repository and return the
```
Small nit: This description says the function returns true and also returns the absolute path to the top level directory when only the latter is true :)

ForgetMe17 commented:
I squashed the commits into one commit.

nishakm commented on tern/analyze/common.py:
```diff
@@ -452,3 +453,76 @@ def get_os_style(image_layer, binary):
+        if isinstance(output, bytes):
+            path_to_toplevel = output.decode('utf-8').split('\n').pop(0)
+    except subprocess.CalledProcessError:
+        logger.error("Cannot find git repo, src path is %s", src_path)
```
Can you use `logger.debug` instead?

nishakm commented on tern/analyze/common.py:
```diff
@@ -452,3 +453,76 @@ def get_os_style(image_layer, binary):
+        if isinstance(output, bytes):
+            sha_info = output.decode('utf-8').split('\n').pop(0)
+    except subprocess.CalledProcessError:
+        logger.error("Cannot find git repo sha, toplevel path is %s",
```
Can you use `logger.debug` instead?

nishakm commented on tern/analyze/common.py:
```diff
@@ -452,3 +453,76 @@ def get_os_style(image_layer, binary):
+    return sha_info
+
+
+def get_git_projcet_name(path_to_toplevel):
```
Change to `get_git_project_name`

nishakm commented on tern/analyze/common.py:
```diff
@@ -452,3 +453,76 @@ def get_os_style(image_layer, binary):
+
+def get_git_projcet_name(path_to_toplevel):
+    '''Given a absolute path to a git repository, return the project name.'''
+    command = ['basename', path_to_toplevel]
```
You can use `os.path.basename` instead of invoking a shell command.

rnjudge commented on tern/analyze/common.py:
```diff
@@ -452,3 +453,76 @@ def get_os_style(image_layer, binary):
+    # get the path of the folder containing the dockerfile
+    dockerfile_folder_path = os.path.dirname(dockerfile_path)
+    # get the path of the target folder or file
+    src_path = os.path.join(dockerfile_folder_path, src_path)
```
This `src_path` is incorrect here. You will want to use `get_git_toplevel(dockerfile_path)` for the `src_path` here and then you don't need to grab it later in line 472.

ForgetMe17 replied:
I am considering this situation:
- dockerfile_folder
    - Dockerfile
    - git_project_folder

ADD command is like `ADD git_project_folder /tmp`. 

In this situation, args given to `check_git_src` equals: 
`src_path = git_project_folder`
`dockerfile_path = path_to_dockerfile_folder/dockerfile_folder/Dockerfile`.

So I use `os.path.join` to get the path `path_to_dockerfile_folder/dockerfile_folder/git_project_folder`.
If `src_path = get_git_toplevel(dockerfile_folder_path)`, we will get the toplevel directory containing this dockerfile instead of looking into the src given in ADD command first.

In another way, if ADD command is copying a folder which contains a git project(assume the name is A) into the image, and the dockerfile itself is in a git repo(assume the name is B), should we record the A or B?
- B
    - subfolder
        - Dockerfile
        - A(src of the ADD command)

rnjudge replied:
> I am considering this situation:
> 
>     * dockerfile_folder
>       
>       * Dockerfile
>       * git_project_folder
> 

Your code works for this situation, but this is a corner case. See below. 

> In another way, if ADD command is copying a folder which contains a git project(assume the name is A) into the image, and the dockerfile itself is in a git repo(assume the name is B), should we record the A or B?
> 
>     * B
>       
>       * subfolder
>         
>         * Dockerfile
>         * A(src of the ADD command)

We want to record the git information from the repo that is being copied, so we would want to record A. However, I think we should focus this PR for now on the following use case and we can add more later.

The most common use case we will see for this is:
* git_project_folder
    * directroy1
    * directory2/Dockerfile

ADD git_project_folder /tmp

which means that with your code now:
`dockerfile_folder_path = ~/git_project_folder/directory2`
`src_path = git_project_folder`
and thus, `src_path = ~/git_project_folder/directory2/git_project_folder`. If you check to see if this directory exists, it will return false and the comment will say "Unable to find the src file or filter".

Because the above is the most common case, I think we should just use `get_top_dir` of the dockerfile location and worry about the other corner cases later. We have a release next week and ideally, this functionality will be included. What do you think?

I also want to apologize, because I think I may have confused you a bit considering all of the situations. I am sorry for that and want you to know that you are doing an excellent job of contributing and making meaningful changes. Your efforts do not go unnoticed!

ForgetMe17 replied:
Thinks for your explanation! Maybe I should figure out the use case clearly before coding. I am enjoying contributing and thanks for encouragement! I will work on it now!

rnjudge commented on tern/analyze/docker/dockerfile.py:
```diff
@@ -265,3 +266,38 @@ def get_base_image_tag(base_instructions):
+            comment_line = find_git_info(command_dict['content'],
+                                         dockerfile_path)
+            dfobj.structure[i]['content'] = command_dict['content'].strip('\n')\
+                + ' # ' + comment_line + '\n'
```
You are updating the `dfobj.structure[i]['content']` correctly here, but we also need to update `dfobj.structure[i]['value']` in the structure dictionary to match. Note that the `value` is the same as the associated `content` value minus the Dockerfile command and newline character. Example of the dictionary from a run command:

```{'instruction': 'RUN', 'startline': 12, 'endline': 12, 'content': 'RUN pip3 install --upgrade pip && pip3 install tern\n', 'value': 'pip3 install --upgrade pip && pip3 install tern'}```

ForgetMe17 commented on tern/analyze/common.py:
```diff
@@ -452,3 +453,61 @@ def get_os_style(image_layer, binary):
+    # get the path of the target folder or file
+    logger.debug('looking into path: %s for git repo.', path_to_toplevel)
+    # add this line to avoid linting check
+    logger.debug('src_path of ADD command: %s.', src_path)
```
I find the `src_path` of ADD command is not used when looking for the toplevel folder. So I add this line to avoid linting check. `src_path` may be useful when considering other corner situations.

nishakm replied:
So if you remove `src_path` these two lines can be removed as well.

rnjudge commented on tern/analyze/common.py:
```diff
@@ -452,3 +453,61 @@ def get_os_style(image_layer, binary):
+    # get the path of the folder containing the dockerfile
+    dockerfile_folder_path = os.path.dirname(dockerfile_path)
+    # locate the top level directory
+    path_to_toplevel = get_git_toplevel(dockerfile_folder_path)
```
What happens if we just do `get_git_toplevel(dockerfile_path)`? Shouldn't we get the same thing? If so, this would eliminate having to create a `dockerfile_folder_path` variable that is only used once.

ForgetMe17 replied:
I set the `cwd(current wording diretory)` to be `dockerfile_folder_path` in `subprocess.check_output` when looking for toplevel dir. `dockerfile_path` points to a file, not a diretory which cannot be set as `cwd`.

nishakm commented on tern/analyze/common.py:
```diff
@@ -452,3 +453,61 @@ def get_os_style(image_layer, binary):
                     os_list=image_layer.os_guess), 'info'))
+
+
+def check_git_src(dockerfile_path, src_path):
```
Seems to me you don't need `src_path` here. You can get the path to the git repository from the directory where the Dockerfile is.

ForgetMe17 replied:
Do we still need to parse `ADD` command to get the `src_path`? since we don't need `src_path` in this function.
I am thinking if `find_git_info` in tern/analyze/docker/dockerfile.py is necessary, because it seems that we don't need to know what is being copied by `ADD` command, and what we use is the `dockerfile path` which is given by a `dfobj`.

nishakm replied:
So it looks to me like `find_git_info` does the ADD command parsing, and you want to check if what got added is a git repository in itself. In this case, you will need functionality to:
1. See if src_path exists (if os.path.exists(src_path))
2. If it does, see if it is a git repository (get_git_toplevel(src_path))
This is only useful if you can produce a full repository url for the git project, otherwise you just get whatever was in the ADD command.
So I think in this case, you do not have to worry about parsing the ADD command to get the `src_path`. We just want to see if the Dockerfile lives in a github repo.

ForgetMe17 replied:
Maybe i can keep the current `find_git_info` which implements parsing ADD command, and not use the parsed `src_path`?

nishakm replied:
Yup! In fact, you don't need to parse it out right now. You can work on that part for #585.

nishakm commented on tern/analyze/common.py:
```diff
@@ -452,3 +453,61 @@ def get_os_style(image_layer, binary):
+    path_to_toplevel = get_git_toplevel(dockerfile_folder_path)
+    # get the path of the target folder or file
+    logger.debug('looking into path: %s for git repo.', path_to_toplevel)
+    # add this line to avoid linting check
```
The reason why the linter was complaining is because you don't use `src_path` anywhere.

nishakm commented on tern/analyze/common.py:
```diff
@@ -452,3 +453,61 @@ def get_os_style(image_layer, binary):
+    # add this line to avoid linting check
+    logger.debug('src_path of ADD command: %s.', src_path)
+    comment_line = ''
+    if os.path.exists(path_to_toplevel):
```
Rather than check if the path exists you can check if `path_to_toplevel` is an empty string just by saying `if path_to_toplevel:`

ForgetMe17 commented:
Thanks for your suggestions!

rnjudge commented on tern/analyze/common.py:
```diff
@@ -452,3 +453,59 @@ def get_os_style(image_layer, binary):
+
+
+def get_git_toplevel(path):
+    '''Give a path, return the absolute path to the top level directory if it
```
small nit that doesn't need to be fixed to merge this but should be noted in the future: "Given" not Give and should note that this has to be a path to a directory, not a file.
